dissect/disc/disc.py

```diff
@@ -0,0 +1,756 @@
+    NotAFileError,
+    NotASymlinkError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ISOFormats(Enum):
+    PLAIN = "plain"
+    ROCKRIDGE = "rockridge"
+    JOLIET = "joliet"
+
+
+DEFAULT_FORMAT_PREFERENCE_ORDER = [
+    ISOFormats.ROCKRIDGE,
+    ISOFormats.JOLIET,
+    ISOFormats.PLAIN,
+]
+
+
+class DISC:
+    """Filesystem implementation for filesystems commonly encountered on optical discs.
+
+    Currently supports ISO9660 and its common extensions Joliet and Rockridge.
+    Not supported: UDF, Apple extensions of ISO9660.
+
+    References:
+
+        ISO9660:
+            - http://www.idea2ic.com/File_Formats/iso9660.pdf
+            - https://wiki.osdev.org/ISO_9660
+            - https://pismotec.com/cfs/iso9660-1999.html
+
+        Rockridge:
+            - https://github.com/torvalds/linux/blob/master/fs/isofs/rock.h
+            - https://studylib.net/doc/18849173/ieee-p1282-rock-ridge-interchange-protocol-draft
+            - https://docplayer.net/29621206-Ieee-p1281-system-use-sharing-protocol-draft-standard-version-1-12-adopted.html
+
+        Joliet:
+            - https://github.com/torvalds/linux/blob/master/fs/isofs/joliet.c
+            - http://littlesvr.ca/isomaster/resources/JolietSpecification.html
+
+        UDF:
+            - https://docplayer.net/237090850-Ecma-tr-universal-disk-format-udf-specification-part-3-revision-2-50-1st-edition-december-reference-number-ecma-123-2009.html
+
+    """  # noqa: E501
+
+    def __init__(self, fh: BinaryIO, preference: Optional[ISOFormats] = None) -> None:
+        self.fh = fh
+        self.iso_format = ISOFormats.PLAIN
+        self.name_encoding = "utf-8"
+
+        self.volume_descriptors = []
+        self._path_table: dict = None
+
+        self.primary_volumes = {
+            ISOFormats.PLAIN: None,
+            ISOFormats.JOLIET: None,
+            ISOFormats.ROCKRIDGE: None,
+        }
+
+        self._load_volume_descriptors()
+
+        if self.primary_volumes[ISOFormats.PLAIN] is None:
+            raise ValueError("No primary volume descriptor found.")
+
+        # To check for rockridge, we need to start parsing the disc as a plain ISO9660 one and then look for extended
+        # attributes.
+        self.logical_block_size = self.primary_volume.logical_block_size
+        self._check_rockridge()
+
+        # At this point we know with which standards this disc is compatible. We can now choose to treat the DISC based
+        # on the preference variable, and we will fall back to another format if the preference is not available.
+        self._select_format(preference)
+
+        # The volume name is padded with spaces, therefore we take everything before the first 'space' character.
+        self.volume_name = self.primary_volume.volume_id.decode(self.name_encoding).split()[0]
+
+    def _load_volume_descriptors(self) -> None:
+        """Load volume descriptors from the ISO file."""
+        self.fh.seek(c_iso.SYSTEM_AREA_SIZE)
+
+        while True:
```

Comment:
If I accidentally throw in something that's not an ISO file, will this read through the entire file? Is there no (sensible) upper limit on how far into the file to read?